src/main/java/com/minecolonies/coremod/colony/crafting/ItemStorageFactory.java

```diff
@@ -80,6 +80,8 @@ public ItemStorage deserialize(IFactoryController controller, PacketBuffer buffe
     {
         final ItemStack stack = buffer.readItemStack();
         final int size = buffer.readVarInt();
+        // PacketBuffer.readItemStack uses a byte internally for item count, and this is limited (to -128 - 127). Override with the ItemStorage value to keep the two values synced, and support larger stacks.
+        stack.setCount(size);
         return this.getNewInstance(stack, size);
```

Comment:
Do we need to set this here to the stack? I thought having it in the storage should be good enough?

Reply:
I don't know.

The short-term problem is that any ItemStorage containing an ItemStack with a number above 127 will be turned into a negative number during serialization sent, and then post-deserialization any ItemStack.get() or ItemStorage.get() will end up returning an Items.AIR object due to how ItemStacks work internally.

The obvious answer is Don't Do That, and never set the count for an ItemStack that's going to get wrapped into an ItemStorage, and that's certainly a viable option for the research loader and packets.  If there's semantic meaning toward ItemStorage.ItemStack.getCount, it's definitely something I can do instead.

It just seems a landmine.

Reply:
The Right Way™ is to assume (and perhaps even ensure) that the ItemStack of an ItemStorage always has a count of 1 (or an otherwise-don't-care number between 1 and 64, anyway).  This does require that anything using ItemStorage.getItemStack() is aware of this and knows not to use the count of the stack for any purpose.  (In general, any code that calls that should be treated as suspicious until proven otherwise, rather than the reverse; since at best it represents a lossy conversion.)

But this is the sort of thing that most storage mods do -- the ItemStack is just a representative item(+nbt) for display/comparison purposes and the actual count is stored elsewhere.

Reply:
You do have to ensure it: once an ItemStack's got a negative count, pretty much anything that uses getItem() or isEmpty(), including isItemStackEqual (isItemStacksEqual will work, but give false positives) and getDisplayName.

At least against accidental behaviors, just changing the ItemStorage constructors will probably be enough for it to never show up, but that seems more Wrong than fixing it here; it's genuinely a serialization-specific issue.

Reply:
It's illegal to set the count of an ItemStack > 64 in the first place.  The real bug is anywhere that sets that.  Assuming that all such bugs are fixed, then there is no need for the serialization code to worry about it, since it's an invalid precondition.

If you really want to write some defensive code, then the right way around is to make the serializer always set the count to 1 before serializing, not to make the deserializer set the stack to a potentially illegal state (which is what this change does).  But (absent other bugs), even that should not be needed.

Reply:
If that’s the case, should I revise the WindowListRecipes view, and how should we handle cases where a recipe input has a count above 64?

Reply:
It's not really possible for recipe inputs to have high counts.  Standard vanilla ingredients only support a single item in each crafting grid cell, so with a 3x3 grid it's impossible to have any single item with a count > 9.

We do have a `CountedIngredient` that does permit vanilla recipes with higher counts (currently max is 32 for compost input), and `CustomRecipe` of course could have higher counts as well, but currently nothing does.  Other mods could do something similar, but again it seems unlikely that they'd do that to a point where it causes an issue.  So I don't think we need to worry about it.

Reply:
Having said that, `RecipeStorage.getCleanedInput` (which is what is shown in the recipe list) does already return ItemStorage, so anything that reads that should be reading the count from the storage, not the stack.

Reply:
I’m not sure there are many places ItemStorage are compared to vanilla recipes.  Maybe the custom recipe must-exist tag?  But that has its own semantic meaning warning that it must follow vanilla rules.  

We don’t use many places where very large ItemStacks make sense, yet (though on spots like the Builder resource tree, that gets closer than you’d think).  But once we open Custom Recipes up for mod pack designer use, it’s both possible and likely to see an outside recipe that wants 1k coal or redstone as an input.  I think such a recipe would even _work_ right now; it just wouldn’t display properly on a remote client.  And short of making a ton on GUI-only copies, I’m not sure how you avoid problems there — does your JEI work handle ItemStorages for such a case, for example?

Reply:
The entire recipe-teaching system revolves around turning vanilla crafting recipes into RecipeStorage, which uses ItemStorage internally.

1k coal in a crafterrecipe?  Possible, perhaps.  Likely, no.  The current implementation of the recipe list will work, but only if the inputs are broken into individual not-more-than-max-stack stacks (such that you could actually fit them into a crafting grid).  It would need some changes if we did want to support recipes with higher counts in a single stack.  But I'm unconvinced that'd be needed in reality.

The current JEI code is outdated, it was written before ItemStorage was used in a lot more places in RecipeStorage, so it's not really representative any more.  Having said that, there's no reason it can't convert an ItemStorage to an ItemStack at the last possible rendering moment, as noted above, especially since it's all client-side anyway.  That's essentially what it does do since (unlike the recipe list) it displays the cleaned and merged inputs, not the original pattern.

Reply:
I mean, on the output side
![image](https://user-images.githubusercontent.com/39036909/116106639-4c3ca680-a680-11eb-97e0-7736b6a258e5.png)
are valid datapack-style recipes.  They're not a _good idea_, but they both display and craft, just having to be split into normally-sized stacks when getting into the player inventory. 

If we want to stick to the philosophy that ItemStacks should always be valid as inventory ItemStacks, unless the second they're being rendered to the GUI, it's doable (though the way List DataProviders are called makes that a bit more expensive than it seems at first glance).

Reply:
yes @uecasm is right here. The count of an itemstack in the itemstorage should never be important. We could go as far as always setting the count of a stack in storage to 1, to make sure there is never a high number in there by any chance. At the moment, if someone gives the storage a stack with > 64 that is a problem right there. If you want it to store a big itemstorage then create the itemstorage with a stack and then set the count.

Reply:
Ok, I'm going to suggest pulling this part of the change. This is really not a serialization problem, this is a problem for ItemStorage itself to solve. 

Talking to Ray, I think we believe the right answer is to update ItemStorage to force the stored stack to 1, and add a new method getStacks() that returns a list of generated/copied stacks that the total equals the amount in the ItemStorage.

Reply:
I think we should stick with the vanilla limitations of never having more than 64 of anything in a valid inventory. If there is more (let's say 128 items research cost) maybe we should render 2 stacks of 64. In that sense, an itemstorage could have a utility method with getStacks that returns a list of stacks with the correct total size.

Reply:
I'm not sure we need the getStacks() method, as long as it's well-enough documented that the getItemStack() _isn't_ a representation of the full containing object.  I'll just remove the deserialization change, and everything research-related use the `GlobalResearch.hasEnoughResources` method (and possibly add a warning in InventoryUtils.getItemCountInItemHandler).

Do you want me to make modifications for cases like the WindowListRecipes et all, or would you rather handle that?

Reply:
Yes, getStack() is solely to get the "stack data" you could say.